### ISSUE TITLE
PathFinding, mutli BlockedByActor check aware

### ISSUE THREAD

**Issue by anvilvapre** [OpenRA/OpenRA#19221]
More of a question then a feature request.

Reading the path finding and `Move.OnFirstRun` activity code it seems that path finding is executed multiple times for multiple `BlockedByActor` states. If no path for the first state is found, the path finding is repeated.

While `BlockedByActor.All` is searched first - it seems it only results in a path in 7% of cases while `None` results in 93% of the cases in a path with a length greater that zero.

BlockedByActor.None 29231
BlockedByActor.Immovable 144
BlockedByActor.All 2221
Total OnFirstrun 31596.

Wouldn't it be potentially be a good optimization to let the path finder be able to search all states at once. This will likely will be slower and would require a cell layer per state. But wouldn't it at least be possible for the the second path search to at least reuse all found partial paths - the cell layer - (bidirectionally both ends) and use that a a starting point? 

I.e. assuming a long path from a to b is blocked in the mid by an Stationary x. The `BlockedByActor.All` would fail. Couldn't the 'BlockedByActor.None' reuse most found path cells?

Why is the path `BlockedByActor.None` still tried to be found after `All`? Can the actor use the blocked by `None` path? Wouldn't it get blocked trying to use/walk it?

**Comment by reaperrr** [comment]
Ping @teinarss

**Comment by anvilvapre** [comment]
@RoosterDragon (while you are working on path finding, you might know some of the answers)

**Comment by RoosterDragon** [comment]
Would be interested in knowing how to repro those figures - my intuition would be most maps and paths would be accessible, unless you've specifically got some units/buildings/trees dividing the map up. Maybe that's a bad intuition and I need to correct it.

> Why is the path BlockedByActor.None still tried to be found after All? Can the actor use the blocked by None path? Wouldn't it get blocked trying to use/walk it?

Correct. I believe the intention is to try and move a unit as far as possible rather than not moving at all, even if you can't actually make it the whole way. The main motivation is to work around poor behaviours moving large blobs of units who run into each other. See #17531 and various related issues/PRs linked to it. (Maybe this is what you need to repro the numbers?)

As for making the pathfinder deal with it better rather than running seperate searches. I think the easiest option would be to use a custom cost. Custom costs are added on top of the usual pathfinding cost for a cell. Let's make up a number larger than any path will cost and call it `n`. When you need the custom cost of a cell, check the blockers and set the custom cost as follows:
- All: `0n`
- Immovable: `1n`
- Stationary: `2n`
- None: `3n`

When you do a search, if a cell is impassable to `All` but passable to `Immovable`, the cell will be added to the open list but with a high cost. It will therefore be at the lower priority in the list than any cells passable via `All` - we know this because we set our `n` to be higher than any path could cost by definition. If a path fails to be found using cells passable to `All`, then all the `Immovable` cells at the bottom of the open list now become candidates and the search will resume from those. The same logic applies similarly to `Stationary` and `None`.

This means a single path search will naturally search via `All` and exhaust the route options for that first. Only when those run out will it then fall back to thinking about `Immovable`, etc, etc. It remembers everything it did so far so we don't end up throwing away any work.

This would be a (likely minor) perf hit compared to a regular search because we'll be shoving a bunch of things into the open list we'd usually be closing instead, but being one search I would expect this to easily pay off compared to running the 2-4 searches you need to run if the first one fails currently.

There's also a practical concern for ensuring `n` is big enough so the above logic works - but small enough so we don't overflow `int` (or alternatively, we may need to bump path costs to `long`).

**Comment by anvilvapre** [comment]
> Would be interested in knowing how to repro those figures - my intuition would be most maps and paths would be accessible, unless you've specifically got some units/buildings/trees dividing the map up. Maybe that's a bad intuition and I need to correct it.

that is a good question, lost my git stash. i assume i simply logged the number of queries per type in `Move.OnFirstRun`. possibly tick time, but i doubt that. i guess i choose the 'Move' activity because i noticed is is used a lot. i typically would run a 3 minute games with 6 or 8 bots. so that would probably mostly be harvesters. 

> Correct. I believe the intention is to try and move a unit as far as possible rather than not moving at all, even if you can't actually make it the whole way. 

ah, that makes sense. otherwise the actor would not move at all. might as well head that way.

regarding the solution you propose, i think i understand what you are saying.

i believe that back then i had a simpler model in mind, where i thought we could reuse the last cell layer/visited cells/partial path to continue the next search. at the risk it might get blocked again - after which it would restart a full search or would perhaps be able to restart the search from the second last cell "search in reverse". but this idea probably doesn't work because if a path search fails the partial path is probably not a single straight path - i.e. one semi path -  but a wide spread area of cells of last visited cells - that all lead to a dead end (?). i.e. for the none path, any one of the previous paths is fine to reuse. 

i could imagine that both your proposal and my idea both only work well if the path is relatively short - where not many cells are touched. with long paths it might be more efficient to restart a full search. just a guess.

**Comment by RoosterDragon** [comment]
Putting it into the custom cost gives you all of that reuse for free. You're only doing one search and it remembers all the costs found so far. The trick is that instead of marking cells are completely inaccessible and thus having to start a fresh search when we run out of candidates, we just give them a higher cost so they're always candidates just of a lower priority. This way they'll get explored eventually, but not until we explore the ones we want to first.

It should be better than starting a new search, because you always get to resume from wherever you left off.

**Comment by anvilvapre** [comment]
@RoosterDragon reading this back. to what extent is this now already covereded in/by hpf? worth gathering some statistics on. or close this?

**Comment by RoosterDragon** [comment]
This isn't covered by HPF, and remains an option to think about.

**Comment by dnqbob** [comment]
https://github.com/OpenRA/OpenRA/issues/20948 is related with this.

If someone think the pathfinding 4 times to a single cell on initializing a `Move` is not horrible enough, then the `MoveWithinRange` and `MoveAdjacentTo` will show you pathfinding 4 times to a bunch of cells in a single tick to initialize a `Move` .

That is the reason of why Attack/Harvest/Capture activity is that costy.
